transformers/modeling_unilm.py

```diff
@@ -0,0 +1,739 @@
+# coding=utf-8
+# Copyright 2018 The Google AI Language Team Authors and The HuggingFace Inc. team.
+# Copyright (c) 2018, NVIDIA CORPORATION.  All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""PyTorch UniLM model. """
+
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+import os
+import sys
+import copy
+import math
+import logging
+import tarfile
+import tempfile
+import shutil
+import numpy as np
+
+import torch
+from torch import nn
+from torch.nn import CrossEntropyLoss, MSELoss
+import torch.nn.functional as F
+from torch.nn.modules.loss import _Loss
+
+from .file_utils import cached_path
+from .modeling_utils import PreTrainedModel
+from .configuration_unilm import UnilmConfig
+from .modeling_bert import ACT2FN, load_tf_weights_in_bert, BertPooler, BertIntermediate, BertOutput, BertPredictionHeadTransform, BertSelfOutput, BertLMPredictionHead, BertOnlyMLMHead, BertOnlyMLMHead, BertEmbeddings
+
+logger = logging.getLogger(__name__)
+
+UNILM_PRETRAINED_MODEL_ARCHIVE_MAP = {
+    'unilm-large-cased': "http://40.85.212.169:8000//pytorch_model.bin"
+}
+
+BertLayerNorm = torch.nn.LayerNorm
+
+
+class BertSelfAttention(nn.Module):
```

Comment:
What is the difference between `BertSelfAttention` defined here and the once defined in `modeling_bert.py`?

Reply:
The forward() method has the argument `history_states` so that we can incrementally perform decoding without re-computing the hidden states of previous time steps.